tika-parsers/src/main/resources/org/apache/tika/parser/captioning/tf/caption_generator.py

```diff
@@ -0,0 +1,185 @@
+        # model state after generating the previous word
+        self.state = state
+        # log probability of the caption
+        self.logprob = logprob
+        # score of the caption
+        self.score = score
+        # optional metadata associated with the partial sentence, if not None, a list of words with the same length
+        self.metadata = metadata
+
+    def __cmp__(self, other):
+        """Compares Captions by score"""
+
+        assert isinstance(other, Caption)
+        if self.score == other.score:
+            return 0
+        elif self.score < other.score:
+            return -1
+        else:
+            return 1
+
+    # for Python 3 compatibility (__cmp__ is deprecated).
+    def __lt__(self, other):
+        assert isinstance(other, Caption)
+        return self.score < other.score
+
+    # also for Python 3 compatibility.
+    def __eq__(self, other):
+        assert isinstance(other, Caption)
+        return self.score == other.score
+
+
+class TopN(object):
+    """Maintains the top N elements of an incrementally provided set"""
+
+    def __init__(self, n):
+        self._n = n
+        self._data = []
+
+    def size(self):
+        assert self._data is not None
+        return len(self._data)
+
+    def push(self, x):
+        """Pushes a new element"""
+
+        assert self._data is not None
+        if len(self._data) < self._n:
+            heapq.heappush(self._data, x)
+        else:
+            heapq.heappushpop(self._data, x)
+
+    def extract(self, sort=False):
+        """
+            Extracts all elements from the TopN. This is a destructive operation,
+            The only method that can be called immediately after extract() is reset()
+        """
+        assert self._data is not None
+        data = self._data
+        self._data = None
+        if sort:
+            data.sort(reverse=True)
+        return data
+
+    def reset(self):
+        """Returns the TopN to an empty state"""
+
+        self._data = []
+
+
+class CaptionGenerator(object):
+    """
+        Class to generate captions from an image-to-text model
+    """
+
+    def __init__(self,
+                 model,
+                 vocab,
+                 beam_size,
+                 max_caption_length,
+                 length_normalization_factor=0.0):
+
+        self.vocab = vocab
+        self.model = model
+
+        self.beam_size = beam_size
+        self.max_caption_length = max_caption_length
+        self.length_normalization_factor = length_normalization_factor
+
+    def beam_search(self, sess, encoded_image):
+        """Runs beam search caption generation on a single image"""
+
+        # feed in the image to get the initial state.
+        initial_state = self.model.feed_image(sess, encoded_image)
+
+        initial_beam = Caption(
+            sentence=[self.vocab.start_id],
+            state=initial_state[0],
+            logprob=0.0,
+            score=0.0,
+            metadata=[""])
```

Comment:
Hey! glad that you pointed this out. metadata is optional. For now it doesn't matter whether it is a list or a dictionary because metadata will always will be null during the process, it is hard coded to be null in the inference step. I was expecting to modify the inference step to assign some useful values as metadata throughout the beam search process. We may or may not need them. Or I can simply remove it, it will speedup the beam search :1st_place_medal: